overlord/managers_test.go

```diff
@@ -9041,6 +9041,295 @@ volumes:
+
+	// let's try updating the kernel;
+	affected, tasksets, err = snapstate.UpdateMany(context.TODO(), st, []string{"pi-kernel"}, 0, &snapstate.Flags{})
+	c.Assert(err, IsNil)
+	sort.Strings(affected)
+	c.Check(affected, DeepEquals, []string{"pi-kernel"})
+
+	chg = st.NewChange("upgrade-snaps", "...")
+	addTaskSetsToChange(chg, tasksets)
+
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+	c.Assert(chg.Err(), ErrorMatches, `(?s).*\(gadget does not consume any of the kernel assets needing synced update "pidtbs"\)`)
+
+	// but we can actually perform the full upgrade set if we first refresh
+	// to an intermediate gadget revision which does not declare an update,
+	// but does now reference the kernel assets, to force going through this
+	// revision, declare it uses a transitional epoch 1* (which can read
+	// epoch 0, the default)
+	gadgetSnapYamlIntermediate := `
+name: pi
+version: 1.0
+type: gadget
+epoch: 1*
+`
+	// while the final gadget will have an epoch 1
+	gadgetSnapYamlFinal := `
+name: pi
+version: 1.0
+type: gadget
+epoch: 1
+`
+	// make both revisions available in the fake store
+	ms.mockSnapUpgradeWithFilesWithRev(c, gadgetSnapYamlIntermediate, "3", [][]string{
+		{"meta/gadget.yaml", intermediaryGadgetYaml},
+		{"boot-assets/start.elf", "start.elf rev1"},
+		// the intermediary gadget snap has these files but it doesn't really
+		// mattter since update does not set an edition, so no update is
+		// attempted using these files
+		{"bcm2710-rpi-2-b.dtb", "bcm2710-rpi-2-b.dtb rev1"},
+		{"bcm2710-rpi-3-b.dtb", "bcm2710-rpi-3-b.dtb rev1"},
+	})
+	ms.mockSnapUpgradeWithFilesWithRev(c, gadgetSnapYamlFinal, "4", [][]string{
+		{"meta/gadget.yaml", finalDesiredGadgetYaml},
+		{"boot-assets/start.elf", "start.elf rev1"},
+	})
+
+	affected, tasksets, err = snapstate.UpdateMany(context.TODO(), st, []string{"pi"}, 0, &snapstate.Flags{})
+	c.Assert(err, IsNil)
+	sort.Strings(affected)
+	c.Check(affected, DeepEquals, []string{"pi"})
+
+	chg = st.NewChange("upgrade-snaps", "...")
+	addTaskSetsToChange(chg, tasksets)
+
+	st.Unlock()
+	err = ms.o.Settle(settleTimeout)
+	st.Lock()
+	c.Assert(err, IsNil)
+	c.Assert(chg.Err(), IsNil)
+
+	// file content is still unchanged because there is no edition bump
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "bcm2710-rpi-2-b.dtb"),
+		testutil.FileContains, "bcm2710-rpi-2-b.dtb rev0")
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "bcm2710-rpi-3-b.dtb"),
+		testutil.FileContains, "bcm2710-rpi-3-b.dtb rev0")
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "overlays/uart0.dtbo"),
+		testutil.FileAbsent)
+	//  gadget content is not updated because there is no edition update
+	c.Check(filepath.Join(dirs.GlobalRootDir, "/run/mnt/", structureName, "start.elf"),
+		testutil.FileContains, "start.elf rev0")
+
+	// pretend we restarted for the gadget refresh (????)
```

Comment:
Does it make sense that we have to do a reboot here? We aren't doing any asset updates so I was confused why the test hangs here if I didn't mock a reboot here and it would also be unexpected in reality if a gadget snap refresh like this did actually require a reboot because nothing except the yaml should have changed

Reply:
ahhhhh yes that makes sense good catch